### PR TITLE
Correct global phase of TwoQubitBasisDecomposer output.

### PR DESCRIPTION
**Pull request by kdk**
### Summary

Correct the global phase of circuits synthesized by the `TwoQubitBasisDecomposer` by simulating the output unitary and applying any needed phase offset. 

on hold for #5474 . 

### Details and comments

Calculating the phase from the input unitary and basis gate unitary should also be possible (and faster), but likely needs correct global phase from #5474. Calculating the phase offset from the operator is also correct but is ~5-10% slower.

### CODE REVIEW

**Comment by levbishop** [comment]
I *think* the 2-qubit part of the decompositions should already maintain phase if the basis and target are both in SU(4), ie. determinant 1. (I'm not sure if I checked it for the python implementation, but this was definitely true for the mathematica version where I developed the decompositions). The only place where a global phase gets changed is on line 110 where a phase is applied to make the determinant=1.

I don't love the approach in this PR partly because of the speed impact mentioned, but also due to possible loss of precision (the PR looks at the first element of magnitude 1e-8 or larger, which may thus lose quite a few digits of precision if the first element is just epsilon).

**Comment by kdk** [comment]
> I _think_ the 2-qubit part of the decompositions should already maintain phase if the basis and target are both in SU(4), ie. determinant 1. (I'm not sure if I checked it for the python implementation, but this was definitely true for the mathematica version where I developed the decompositions). The only place where a global phase gets changed is on line 110 where a phase is applied to make the determinant=1.
>
> I don't love the approach in this PR partly because of the speed impact mentioned, but also due to possible loss of precision (the PR looks at the first element of magnitude 1e-8 or larger, which may thus lose quite a few digits of precision if the first element is just epsilon).

From what I've seen, this is correct (phase being preserved if both basis and target are SU(4)). But even with #5474, I still have a handful of test failures (for cases where target and or basis were outside SU(4)) that I wasn't able to track down. I'd propose we move this ahead (to resolve #5606 and unblock #5577 ) and I'll fix those cases in a different PR.

File: qiskit/quantum_info/synthesis/two_qubit_decompose.py [hunk 1]
```diff
@@ -442,6 +442,14 @@ def __call__(self, target, basis_fidelity=None):
         return_circuit.compose(decomposition_euler[2*best_nbasis], [q[0]], inplace=True)
         return_circuit.compose(decomposition_euler[2*best_nbasis+1], [q[1]], inplace=True)
 
+        return_operator = Operator(return_circuit).data
+        for idx, elt in np.ndenumerate(return_operator):
+            if abs(elt) > 1e-8:
```
Comment (mtreinish) [inline qiskit/quantum_info/synthesis/two_qubit_decompose.py:447]: Is there a reason you went for `1e-8` here? The normal default tolerance we use elsewhere are on the order of `1e-15`, `1e-12`, and `1e-13` for example:

https://github.com/Qiskit/qiskit-terra/blob/b9027930e014acc822798ff68baca9b5564c85bd/qiskit/quantum_info/synthesis/two_qubit_decompose.py#L95
https://github.com/Qiskit/qiskit-terra/blob/b9027930e014acc822798ff68baca9b5564c85bd/qiskit/quantum_info/synthesis/two_qubit_decompose.py#L41
https://github.com/Qiskit/qiskit-terra/blob/b9027930e014acc822798ff68baca9b5564c85bd/qiskit/quantum_info/synthesis/two_qubit_decompose.py#L69

Comment (kdk) [reply qiskit/quantum_info/synthesis/two_qubit_decompose.py:447]: This is a good question. The approach and tolerances here are drawn from what we use for `Operator.equiv` ( https://github.com/Qiskit/qiskit-terra/blob/a0e7dcec145db9bd3ef7e5bf1566c23292db31a1/qiskit/quantum_info/operators/predicates.py#L64 ).

To @levbishop 's comment about preserving precision, I think we'd want the tolerance to be larger to decrease the likelihood we find an element that's order epsilon. We could scan the `np.abs` of the full array, and use the largest element from there, or maybe increase the tolerance to the smallest possible value that can be the largest magnitude in a unitary matrix, which I think is at least 1/n<sup>2</sup>, but maybe could be higher.

**Comment by kdk** [comment]
Closing in favor of #5764 .

**Pull request closed without merging**